lib/mix/test/mix/tasks/compile.erlang_test.exs

```diff
@@ -87,10 +88,43 @@ defmodule Mix.Tasks.Compile.ErlangTest do
       -export([version/0]).
       version() -> v2.
       """
-      assert Mix.Tasks.Compile.Erlang.run([]) == :ok
+      assert Mix.Tasks.Compile.Erlang.run([]) == {:ok, []}
 
       # If the module was not purged on recompilation, this would fail.
       assert :v2 == :purge_test.version
     end
   end
+
+  test "saves warnings between builds" do
+    in_fixture "compile_erlang", fn ->
+      file = Path.absname("src/has_warning.erl")
+      File.write! file, """
+      -module(has_warning).
+      my_fn() -> ok.
+      """
+
+      capture_io fn ->
+        assert {:ok, [%Mix.Task.Compiler.Diagnostic{
+          file: ^file,
+          compiler_name: "erl_lint",
+          message: "function my_fn/0 is unused",
+          position: 2,
+          severity: :warning
+        } = warning]} = Mix.Tasks.Compile.Erlang.run([])
+
+        # Should return warning without recompiling file
```

Comment:
Should we also support the `--all-warnings` option? It would probably be nice to make everything consistent.